File: Sources/Drivers/Thrustmaster/TMCommon/T300FFBEncoder.swift
import WheelProtocol

public enum T300FFBEncoder {

    public static let opcodeConstant:  UInt8 = 0x6A
    public static let opcodeCondition: UInt8 = 0x64
    public static let opcodePeriodic:  UInt8 = 0x6B
    public static let opcodePlay:      UInt8 = 0x89
    public static let codePlay:        UInt8 = 0x41

    private static func waveformByte(_ kind: EffectKind) -> UInt8? {
        switch kind {
        case .squarePeriodic:       return 0x01
        case .trianglePeriodic:     return 0x02
        case .sinePeriodic:         return 0x03
        case .sawtoothUpPeriodic:   return 0x04
        case .sawtoothDownPeriodic: return 0x05
        default:                    return nil
        }
    }

    private static let conditionHardcoded: [UInt8] = [
        0xFE, 0xFF, 0xFE, 0xFF, 0xFE, 0xFF, 0xFE, 0xFF,
    ]

    public static func encode(_ effect: NormalizedEffect) throws -> [USBPacket] {
        switch effect {
        case .constant(let slot, let magnitude, let duration, _, let env):
            return [constantUpload(slot: slot,
                                   magnitude: magnitude,
                                   duration: duration,
                                   envelope: env)]
        case .spring(let slot, let p):
            return [conditionUpload(slot: slot, params: p, kind: .spring)]
        case .damper(let slot, let p):
            return [conditionUpload(slot: slot, params: p, kind: .damper)]
        case .friction(let slot, let p):
            return [conditionUpload(slot: slot, params: p, kind: .friction)]
        case .inertia(let slot, let p):
            return [conditionUpload(slot: slot, params: p, kind: .inertia)]
        case .periodic(let slot, let kind, let params, let duration, let env):
            guard let wave = Self.waveformByte(kind) else { throw DriverError.effectNotSupported(kind) }
            return [periodicUpload(slot: slot,
                                   waveform: wave,
                                   params: params,
                                   duration: duration,
                                   envelope: env)]
        case .ramp, .customForceData:
            throw DriverError.notImplemented
        }
    }

    public static func playPacket(slot: UInt8, repeats: UInt16 = 1) -> USBPacket {
        let count = (repeats == 0 || repeats >= 0xFFFF) ? UInt16(0) : repeats
        let bytes: [UInt8] = [
            0x00, slot &+ 1, opcodePlay,
            codePlay,
            UInt8(count & 0xFF), UInt8(count >> 8),
        ]
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    public static func stopPacket(slot: UInt8) -> USBPacket {
        let bytes: [UInt8] = [0x00, slot &+ 1, opcodePlay, 0x00]
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private enum ConditionKind {
        case spring, damper, friction, inertia
        var typeByte: UInt8 { self == .spring ? 0x06 : 0x07 }
        var maxSaturation: UInt16 { self == .spring ? 0x6AA6 : 0x7FFC }
    }

    private static func constantUpload(slot: UInt8,
                                       magnitude: Int16,
                                       duration: UInt32,
                                       envelope: Envelope?) -> USBPacket
    {
        let level = Int16(max(-16385, min(16381, Int(magnitude) / 2)))
        let dur = duration == 0 ? UInt16(0xFFFF) : UInt16(min(duration, UInt32(UInt16.max - 1)))

        var bytes: [UInt8] = [0x00, slot &+ 1, opcodeConstant]
        bytes += le16(UInt16(bitPattern: level))
        bytes += envelopeBytes(envelope)
        bytes += [0x00]
        bytes += timingBytes(durationMs: dur, offsetMs: 0)
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private static func conditionUpload(slot: UInt8,
                                        params: ConditionParams,
                                        kind: ConditionKind) -> USBPacket
    {
        let rightCoeff = Int16(max(-32767, min(32767, Int(params.positiveCoefficient))))
        let leftCoeff  = Int16(max(-32767, min(32767, Int(params.negativeCoefficient))))
        let halfDead = Int(params.deadBand) / 2
        let rightDead = Int16(max(-0x7FFF, min(0x7FFF, Int(params.centerOffset) + halfDead)))
        let leftDead  = Int16(max(-0x7FFF, min(0x7FFF, Int(params.centerOffset) - halfDead)))
        let rightSat = scaleSaturation(params.positiveSaturation, max: kind.maxSaturation)
        let leftSat  = scaleSaturation(params.negativeSaturation, max: kind.maxSaturation)

        var bytes: [UInt8] = [0x00, slot &+ 1, opcodeCondition]
        bytes += le16(UInt16(bitPattern: rightCoeff))
        bytes += le16(UInt16(bitPattern: leftCoeff))
        bytes += le16(UInt16(bitPattern: rightDead))
        bytes += le16(UInt16(bitPattern: leftDead))
        bytes += le16(rightSat)
        bytes += le16(leftSat)
        bytes += conditionHardcoded
        bytes += le16(kind.maxSaturation)
        bytes += le16(kind.maxSaturation)
        bytes += [kind.typeByte]
        bytes += timingBytes(durationMs: 0xFFFF, offsetMs: 0)
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private static func periodicUpload(slot: UInt8,
                                       waveform: UInt8,
                                       params: PeriodicParams,
                                       duration: UInt32,
                                       envelope: Envelope?) -> USBPacket
    {
        let dur = duration == 0 ? UInt16(0xFFFF) : UInt16(min(duration, UInt32(UInt16.max - 1)))
        var bytes: [UInt8] = [0x00, slot &+ 1, opcodePeriodic]
        bytes += le16(UInt16(bitPattern: params.magnitude))
        bytes += le16(UInt16(bitPattern: params.offset))
        bytes += le16(params.phase)
        bytes += le16(UInt16(min(params.period, UInt32(UInt16.max))))
        bytes += le16(0x8000)
        bytes += envelopeBytes(envelope)
        bytes += [waveform]
        bytes += timingBytes(durationMs: dur, offsetMs: 0)
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private static func scaleSaturation(_ sat: Int16, max: UInt16) -> UInt16 {
        if sat == 0 { return max }
        return UInt16(UInt32(UInt16(bitPattern: sat)) * UInt32(max) / 0xFFFF)
    }

    private static func envelopeBytes(_ env: Envelope?) -> [UInt8] {
        guard let e = env else { return [0, 0, 0, 0, 0, 0, 0, 0] }
        let attackLen = UInt16(min(e.attackTime, UInt32(UInt16.max)))
        let attackLvl = UInt16(bitPattern: e.attackLevel)
        let fadeLen   = UInt16(min(e.fadeTime, UInt32(UInt16.max)))
        let fadeLvl   = UInt16(bitPattern: e.fadeLevel)
        return le16(attackLen) + le16(attackLvl) + le16(fadeLen) + le16(fadeLvl)
    }

    private static func timingBytes(durationMs: UInt16, offsetMs: UInt16) -> [UInt8] {
        var b: [UInt8] = []
        b += [0x4F]
        b += le16(durationMs)
        b += [0x00, 0x00]
        b += le16(offsetMs)
        b += [0x00]
        b += [0xFF, 0xFF]
        return b
    }

    private static func le16(_ v: UInt16) -> [UInt8] {
        [UInt8(v & 0xFF), UInt8(v >> 8)]
    }
}

File: Sources/DEXT/T300FFBEncoder.cpp
#include "T300FFBEncoder.hpp"

namespace T300FFBEncoder {

static constexpr uint8_t kOpcodeConstant  = 0x6A;
static constexpr uint8_t kOpcodeCondition = 0x64;
static constexpr uint8_t kOpcodePeriodic  = 0x6B;
static constexpr uint8_t kOpcodeRamp      = 0x6B;
static constexpr uint8_t kOpcodePlay      = 0x89;
static constexpr uint8_t kCodePlay        = 0x41;

static const uint8_t kConditionHardcoded[8] = {
    0xFE, 0xFF, 0xFE, 0xFF, 0xFE, 0xFF, 0xFE, 0xFF,
};

static void writeLe16(uint8_t *dst, uint16_t v) {
    dst[0] = (uint8_t)(v & 0xFF);
    dst[1] = (uint8_t)(v >> 8);
}

static void writeEnvelope(uint8_t *dst, const Envelope *env) {
    if (!env) {
        for (int i = 0; i < 8; ++i) dst[i] = 0;
        return;
    }
    uint16_t attackLen =
        (env->attackTime > 0xFFFF) ? 0xFFFF : (uint16_t)env->attackTime;
    uint16_t fadeLen =
        (env->fadeTime   > 0xFFFF) ? 0xFFFF : (uint16_t)env->fadeTime;
    writeLe16(dst,     attackLen);
    writeLe16(dst + 2, (uint16_t)env->attackLevel);
    writeLe16(dst + 4, fadeLen);
    writeLe16(dst + 6, (uint16_t)env->fadeLevel);
}

static void writeTiming(uint8_t *dst, uint16_t durationMs, uint16_t offsetMs) {
    dst[0] = 0x4F;
    writeLe16(dst + 1, durationMs);
    dst[3] = 0x00;
    dst[4] = 0x00;
    writeLe16(dst + 5, offsetMs);
    dst[7] = 0x00;
    dst[8] = 0xFF;
    dst[9] = 0xFF;
}

static int32_t clamp32(int32_t v, int32_t lo, int32_t hi) {
    if (v < lo) return lo;
    if (v > hi) return hi;
    return v;
}

static uint16_t scaleSaturation(int16_t sat, uint16_t maxSat) {
    if (sat == 0) return maxSat;
    uint32_t s = (uint32_t)(uint16_t)sat;
    return (uint16_t)(s * (uint32_t)maxSat / 0xFFFF);
}

static uint16_t clampDuration(uint32_t durationMs) {
    if (durationMs == 0) return 0xFFFF;
    if (durationMs > 0xFFFE) return 0xFFFE;
    return (uint16_t)durationMs;
}

size_t constantUploadPacket(
    uint8_t slot, int16_t magnitude, uint32_t durationMs,
    const Envelope *env,
    uint8_t *out, size_t outCap)
{
    if (outCap < 24) return 0;
    int32_t level = clamp32((int32_t)magnitude / 2, -16385, 16381);
    uint16_t dur = clampDuration(durationMs);
    out[0] = 0x00;
    out[1] = (uint8_t)(slot + 1);
    out[2] = kOpcodeConstant;
    writeLe16(out + 3, (uint16_t)(int16_t)level);
    writeEnvelope(out + 5, env);
    out[13] = 0x00;
    writeTiming(out + 14, dur, 0);
    return 24;
}

size_t periodicUploadPacket(
    uint8_t slot, WaveformCode wave,
    int16_t magnitude, int16_t offset,
    uint16_t phase, uint32_t period,
    uint32_t durationMs, const Envelope *env,
    uint8_t *out, size_t outCap)
{
    if (outCap < 32) return 0;
    uint16_t periodU16 = (period > 0xFFFF) ? 0xFFFF : (uint16_t)period;
    uint16_t dur = clampDuration(durationMs);
    out[0] = 0x00;
    out[1] = (uint8_t)(slot + 1);
    out[2] = kOpcodePeriodic;
    writeLe16(out + 3,  (uint16_t)magnitude);
    writeLe16(out + 5,  (uint16_t)offset);
    writeLe16(out + 7,  phase);
    writeLe16(out + 9,  periodU16);
    writeLe16(out + 11, 0x8000);
    writeEnvelope(out + 13, env);
    out[21] = (uint8_t)wave;
    writeTiming(out + 22, dur, 0);
    return 32;
}

size_t rampUploadPacket(
    uint8_t slot, int16_t start, int16_t end,
    uint32_t durationMs, const Envelope *env,
    uint8_t *out, size_t outCap)
{
    if (outCap < 32) return 0;
    int32_t diff = (int32_t)start - (int32_t)end;
    if (diff < 0) diff = -diff;
    uint16_t slope = (uint16_t)(diff / 2);
    int16_t center = (int16_t)(((int32_t)start + (int32_t)end) / 2);
    uint8_t invert = (start < end) ? 0x04 : 0x05;
    uint16_t dur = clampDuration(durationMs);
    out[0] = 0x00;
    out[1] = (uint8_t)(slot + 1);
    out[2] = kOpcodeRamp;
    writeLe16(out + 3,  slope);
    writeLe16(out + 5,  (uint16_t)center);
    out[7]  = 0x00;
    out[8]  = 0x00;
    writeLe16(out + 9,  dur);
    writeLe16(out + 11, 0x8000);
    writeEnvelope(out + 13, env);
    out[21] = invert;
    writeTiming(out + 22, dur, 0);
    return 32;
}

size_t conditionUploadPacket(
    uint8_t slot, ConditionKind kind,
    int16_t posCoeff, int16_t negCoeff,
    int16_t posSat,   int16_t negSat,
    uint16_t deadBand, int16_t centerOffset,
    uint8_t *out, size_t outCap)
{
    if (outCap < 38) return 0;
    uint16_t maxSat   = (kind == CondSpring) ? 0x6AA6 : 0x7FFC;
    uint8_t  typeByte = (kind == CondSpring) ? 0x06   : 0x07;

    int16_t rightCoeff = (int16_t)clamp32(posCoeff, -32767, 32767);
    int16_t leftCoeff  = (int16_t)clamp32(negCoeff, -32767, 32767);
    int32_t halfDead   = (int32_t)deadBand / 2;
    int16_t rightDead  = (int16_t)clamp32((int32_t)centerOffset + halfDead, -32767, 32767);
    int16_t leftDead   = (int16_t)clamp32((int32_t)centerOffset - halfDead, -32767, 32767);
    uint16_t rightSat  = scaleSaturation(posSat, maxSat);
    uint16_t leftSat   = scaleSaturation(negSat, maxSat);

    out[0] = 0x00;
    out[1] = (uint8_t)(slot + 1);
    out[2] = kOpcodeCondition;
    writeLe16(out + 3,  (uint16_t)rightCoeff);
    writeLe16(out + 5,  (uint16_t)leftCoeff);
    writeLe16(out + 7,  (uint16_t)rightDead);
    writeLe16(out + 9,  (uint16_t)leftDead);
    writeLe16(out + 11, rightSat);
    writeLe16(out + 13, leftSat);
    for (int i = 0; i < 8; ++i) out[15 + i] = kConditionHardcoded[i];
    writeLe16(out + 23, maxSat);
    writeLe16(out + 25, maxSat);
    out[27] = typeByte;
    writeTiming(out + 28, 0xFFFF, 0);
    return 38;
}

size_t playPacket(uint8_t slot, uint16_t repeats,
                  uint8_t *out, size_t outCap)
{
    if (outCap < 6) return 0;
    uint16_t count = (repeats == 0 || repeats >= 0xFFFF) ? 0 : repeats;
    out[0] = 0x00;
    out[1] = (uint8_t)(slot + 1);
    out[2] = kOpcodePlay;
    out[3] = kCodePlay;
    writeLe16(out + 4, count);
    return 6;
}

size_t stopPacket(uint8_t slot, uint8_t *out, size_t outCap) {
    if (outCap < 4) return 0;
    out[0] = 0x00;
    out[1] = (uint8_t)(slot + 1);
    out[2] = kOpcodePlay;
    out[3] = 0x00;
    return 4;
}

} // namespace T300FFBEncoder
